### PR TITLE
Add json output option

### LINKED ISSUE #56: output report in a machine-readable format
As far as I've seen, is not possible to output the result in a machine parseable format yet.

For instance, running `flay -# -d lib/*.rb` in this project outputs:

```
$ flay -# -d lib/*.rb
Total score (lower is better) = 72

Similar code found in :iter (mass = 36)
  A: lib/flay.rb:74
  B: lib/flay.rb:99

A: opts.on("-m", "--mass MASS", Integer, "Sets mass threshold (default = #{options[:mass]})") do |m|
B: opts.on("-t", "--timeout TIME", Integer, "Set the timeout. (default = #{options[:timeout]})") do |t|
A:   options[:mass] = m.to_i
B:   options[:timeout] = t.to_i
   end

Similar code found in :defn (mass = 36)
  A: lib/flay_erb.rb:28
  B: lib/flay_erb.rb:36

A: def add_expr_literal(src, code)
B: def add_expr_escaped(src, code)
     if code.=~(BLOCK_EXPR) then
A:     ((src << "@output_buffer.append= ") << code)
B:     ((src << "@output_buffer.safe_append= ") << code)
     else
A:     (((src << "@output_buffer.append=(") << code) << ");")
B:     (((src << "@output_buffer.safe_append=(") << code) << ");")
     end
   end
```

Something like this would be useful for parsing (JSON example, XML or YAML would also be valid):

```
{
	"total": 72,
	"clones": [
		{
			"match": ":iter",
			"mass": 36,
			"A": {
				"filename": "lib/flay.rb",
				"line": 74
			},
			"B": {
				"filename": "lib/flay.rb",
				"line": 99
			},
			"lines": [
				{
					"source": "A",
					"content": "opts.on("-m", "--mass MASS", Integer, "Sets mass threshold (default = #{options[:mass]})") do |m|"
				},
				{
					"source": "B",
					"content": "opts.on("-t", "--timeout TIME", Integer, "Set the timeout. (default = #{options[:timeout]})") do |t|"
				},
				{
					"source": "A",
					"content": "   options[:mass] = m.to_i"
				},
				{
					"source": "B",
					"content": "   options[:timeout] = t.to_i"
				},
				{
					"source": "Common",
					"content": "   end"
				}
			]
		},
		{
			"match": ":defn",
			"mass": 36,
			"A": {
				"filename": "lib/flay_erb",
				"line": 28
			},
			"B": {
				"filename": "lib/flay_erb",
				"line": 36
			},
			"lines": [
				{
					"source": "A",
					"content": "def add_expr_literal(src, code)"
				},
				{
					"source": "B",
					"content": "def add_expr_escaped(src, code)"
				},
				{
					"source": "Common",
					"content": "    if code.=~(BLOCK_EXPR) then"
				},
				{
					"source": "A",
					"content": "     ((src << "@output_buffer.append= ") << code)"
				},
				{
					"source": "B",
					"content": "     ((src << "@output_buffer.safe_append= ") << code)"
				},
				{
					"source": "Common",
					"content": "     else"
				},
				{
					"source": "A",
					"content": "     (((src << "@output_buffer.append=(") << code) << ");")"
				},
				{
					"source": "B",
					"content": "     (((src << "@output_buffer.safe_append=(") << code) << ");")"
				},
				{
					"source": "Common",
					"content": "     end"
				},
				{
					"source": "Common",
					"content": "   end"
				}
			]
		}
	]
}
```

Would anyone oppose to this? If no one opposes I'm willing to start the feature and do the PR

### PR DESCRIPTION
**Pull request by pedrorijo91**
closes #56 

Please let me know if there's something you don't totally agree.

Example output for Rakefile with command `ruby "#{Hoe::RUBY_FLAGS} bin/flay -d -f lib/*.rb"`

```
{"total":114,"clones":[{"prefix":"1","match":"Similar","mass":42,"name":"block","files":[{"filename":"lib/flay.rb","line":550,"contents":["msg = \"sexp_to_#{File.extname(node.file).sub(/./, \"\")}\"\nself.respond_to?(msg) ? (self.send(msg, node)) : (sexp_to_rb(node))\n"]},{"filename":"lib/flay.rb","line":609,"contents":["msg = \"sexp_to_#{File.extname(s.file).sub(/./, \"\")}\"\nself.respond_to?(msg) ? (self.send(msg, s)) : (sexp_to_rb(s))\n"]}]},{"prefix":"2","match":"Similar","mass":36,"name":"iter","files":[{"filename":"lib/flay.rb","line":76,"contents":["opts.on(\"-m\", \"--mass MASS\", Integer, \"Sets mass threshold (default = #{options[:mass]})\") do |m|\n  options[:mass] = m.to_i\nend"]},{"filename":"lib/flay.rb","line":101,"contents":["opts.on(\"-t\", \"--timeout TIME\", Integer, \"Set the timeout. (default = #{options[:timeout]})\") do |t|\n  options[:timeout] = t.to_i\nend"]}]},{"prefix":"3","match":"Similar","mass":36,"name":"defn","files":[{"filename":"lib/flay_erb.rb","line":28,"contents":["def add_expr_literal(src, code)\n  if code.=~(BLOCK_EXPR) then\n    ((src << \"@output_buffer.append= \") << code)\n  else\n    (((src << \"@output_buffer.append=(\") << code) << \");\")\n  end\nend"]},{"filename":"lib/flay_erb.rb","line":36,"contents":["def add_expr_escaped(src, code)\n  if code.=~(BLOCK_EXPR) then\n    ((src << \"@output_buffer.safe_append= \") << code)\n  else\n    (((src << \"@output_buffer.safe_append=(\") << code) << \");\")\n  end\nend"]}]}]}
```

or, after formatted:

```
{
  "total": 114,
  "clones": [
    {
      "prefix": "1",
      "match": "Similar",
      "mass": 42,
      "name": "block",
      "files": [
        {
          "filename": "lib/flay.rb",
          "line": 550,
          "contents": [
            "msg = \"sexp_to_#{File.extname(node.file).sub(/./, \"\")}\"\nself.respond_to?(msg) ? (self.send(msg, node)) : (sexp_to_rb(node))\n"
          ]
        },
        {
          "filename": "lib/flay.rb",
          "line": 609,
          "contents": [
            "msg = \"sexp_to_#{File.extname(s.file).sub(/./, \"\")}\"\nself.respond_to?(msg) ? (self.send(msg, s)) : (sexp_to_rb(s))\n"
          ]
        }
      ]
    },
    {
      "prefix": "2",
      "match": "Similar",
      "mass": 36,
      "name": "iter",
      "files": [
        {
          "filename": "lib/flay.rb",
          "line": 76,
          "contents": [
            "opts.on(\"-m\", \"--mass MASS\", Integer, \"Sets mass threshold (default = #{options[:mass]})\") do |m|\n  options[:mass] = m.to_i\nend"
          ]
        },
        {
          "filename": "lib/flay.rb",
          "line": 101,
          "contents": [
            "opts.on(\"-t\", \"--timeout TIME\", Integer, \"Set the timeout. (default = #{options[:timeout]})\") do |t|\n  options[:timeout] = t.to_i\nend"
          ]
        }
      ]
    },
    {
      "prefix": "3",
      "match": "Similar",
      "mass": 36,
      "name": "defn",
      "files": [
        {
          "filename": "lib/flay_erb.rb",
          "line": 28,
          "contents": [
            "def add_expr_literal(src, code)\n  if code.=~(BLOCK_EXPR) then\n    ((src << \"@output_buffer.append= \") << code)\n  else\n    (((src << \"@output_buffer.append=(\") << code) << \");\")\n  end\nend"
          ]
        },
        {
          "filename": "lib/flay_erb.rb",
          "line": 36,
          "contents": [
            "def add_expr_escaped(src, code)\n  if code.=~(BLOCK_EXPR) then\n    ((src << \"@output_buffer.safe_append= \") << code)\n  else\n    (((src << \"@output_buffer.safe_append=(\") << code) << \");\")\n  end\nend"
          ]
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: lib/flay.rb [hunk 1]
```diff
@@ -498,17 +504,74 @@ def summary
...
+    end
+
+    io.puts json.to_json
+  end
+
+  def report_io io, data
     io.puts "Total score (lower is better) = #{self.total}"
 
-    if option[:summary] then
+    if option[:summary]
```
Comment (zenspider) [inline lib/flay.rb]: Don't reformat my code please.

File: lib/flay.rb [hunk 2]
```diff
@@ -525,7 +588,7 @@ def report io = $stdout
 
       io.puts
       io.puts "%s%s code found in %p (mass%s = %d)" %
-        [prefix, match, item.name, item.bonus, item.mass]
+                  [prefix, match, item.name, item.bonus, item.mass]
```
Comment (zenspider) [inline lib/flay.rb]: Ditto

File: lib/flay.rb [hunk 3]
```diff
@@ -548,6 +611,21 @@ def report io = $stdout
...
+
+  def report io = $stdout
+    only = option[:only]
+
+    data = analyze only
+
+    if option[:report]
+      report_json(io,data)
+    else
+      report_io(io,data)
```
Comment (zenspider) [inline lib/flay.rb]: please follow my coding conventions

File: lib/flay.rb (hunk 1 above)
Comment (pedrorijo91) [inline lib/flay.rb]: you are totally right, didn't notice. Going to fix

File: lib/flay.rb (hunk 2 above)
Comment (pedrorijo91) [inline lib/flay.rb]: same

> you are totally right, didn't notice. Going to fix

File: lib/flay.rb (hunk 3 above)
Comment (pedrorijo91) [inline lib/flay.rb]: not sure what you mean. Didn't see anything about conventions, and I'm not used to ruby and to its best code style practices. If you can specify the problem I will fix it in no time

**Comment by pedrorijo91** [comment]
Hi,

Is there anything you wish me to change? Or is anything else blocking this PR?

**Comment by pedrorijo91** [comment]
@zenspider

**Pull request closed without merging**